File: KiLCA/progs/main_linear.cpp
/*! \file
    \brief The basic driver program to run KiLCA code.
*/

#include <stdlib.h>
#include <stdio.h>
#include <cmath>
#include <cstring>
#include <climits>
#include <iostream>
#include <fstream>
#include <unistd.h>

#include "core.h"
#include "main_linear.h"

using namespace std;

/*******************************************************************/
//!The main program
/*!
Calls KiLCA library functions
*/
int main (int argc, char **argv)
{

//!gets the path to the project from the current dir or the command line
char *path = new char[1024];

if (argc == 1)
{
   getcwd (path, 1024); //current directory
}
else if (argc >= 2)
{
   strcpy (path, argv[1]); //command line argument
}

if (path[strlen(path)-1] != '/') strcat(path, "/");

intptr_t cd = core_data_create_ (path); //!<core data structure contains pointers to all important code data
set_core_data_in_core_module_ (&cd);  //!<stores the core data pointer in fortran module for later use

core_data_calc_and_set_mode_independent_ (cd); //!<evident from the function name

if (get_antenna_flag_eigmode_ () == 0)
{
    core_data_calc_and_set_mode_dependent_antenna_ (cd); //!<normal operation with antenna
}
else
{
    core_data_calc_and_set_mode_dependent_eigmode_ (cd); //!<special operation for instabilities hunting
}

//!frees the used memory
delete [] path;

core_data_destroy_ (cd);

return 0;
}

/*******************************************************************/

File: KiLCA/progs/main_linear.f90
!> The basic driver program to run KiLCA code, ported from progs/main_linear.cpp.
!> Builds the project path (current directory or the first command-line
!> argument), then drives the now-Fortran core-data pipeline: create the core
!> data, register its handle in the core module, compute the mode-independent
!> data, then dispatch to the antenna or eigenmode mode-dependent path.
program main_linear
    use kilca_legacy_interfaces_m, only: set_core_data_in_core_module
    use kilca_antenna_settings_m, only: get_antenna_flag_eigmode
    use, intrinsic :: iso_c_binding, only: c_intptr_t
    use kilca_core_data_m, only: core_data_create_, core_data_destroy_, &
        core_data_calc_and_set_mode_independent_, &
        core_data_calc_and_set_mode_dependent_antenna_, &
        core_data_calc_and_set_mode_dependent_eigmode_
    use kilca_progs_common_m, only: get_project_path, to_cstr
    implicit none
    integer(c_intptr_t) :: cd

    cd = core_data_create_(get_project_path())
    call set_core_data_in_core_module(cd)

    call core_data_calc_and_set_mode_independent_(cd)

    if (get_antenna_flag_eigmode() == 0) then
        call core_data_calc_and_set_mode_dependent_antenna_(cd)
    else
        call core_data_calc_and_set_mode_dependent_eigmode_(cd)
    end if

    call core_data_destroy_(cd)
end program main_linear
